### DISCUSSION TITLE
Working with macros (e.g. Observable)

### DISCUSSION THREAD

**Discussion by hadiidbouk** [pointfreeco/swift-dependencies#99] [General]
With the introduction of `macros` and especially `@Observable`, the library seems incompatible with the `ObservableMacro`, and we need to provide an initial value of our dependencies.

![Screenshot 2023-06-07 at 2 23 57 PM](https://github.com/pointfreeco/swift-dependencies/assets/19314956/78db7ddd-b29c-4630-8175-ceeb581b62ac)

Any thoughts on how we can support the `ObservableMacro`? Should we convert the property wrapper to a `macro`?

**Comment by hadiidbouk** [comment]
A workaround can be doing something like this:

```swift
@Observable
final class MarvelHeroesViewModel {
	var superHeroes: [MarvelCharacter] = []
	var isLoading = false

	let marvelHeroesService: MarvelHeroesService

	init() {
		@Dependency(\.marvelHeroesService) var marvelHeroesService
		self.marvelHeroesService = marvelHeroesService
		self.loadData()
	}
```

**Comment by mbrandonw** [comment]
Hi @hadiidbouk, there are two things happening here: one is definitely a bug and the other is a limitation of `@Observable` that will hopefully someday be lifted.

The bug is that property wrappers currently do not work in `@Observable`. I'm pretty sure this is a bug because even Apple's migration docs have a code snippet that does not currently compile due to the `@Published` property wrapper:

```swift
@Observable 
class Library: ObservableObject {
  @Published var books: [Book] = [Book(), Book(), Book ()]
}
```

And the limitation is that `@Observable` requires all observed fields to have a default value. I'm not sure if this is a necessary limitation or if it could be lifted someday.

However, even with the bug and limitation it is still possible to use `@Dependency` in an observable object. The problem is that the `@Observable` macro is trying to expand all types of code in order to observe changes to the dependency, but we don't need to do that. We can just not observe the dependency:

```swift
@Observable
class Model {
  @ObservationIgnored @Dependency(\.date) var date
}
```

That compiles just fine.

**Comment by Esquilli** [comment]
Requiring a default value is a known bug ([link](https://developer.apple.com/documentation/ios-ipados-release-notes/ios-ipados-17-release-notes#Swift)). So, it should work properly after that is fixed?

![image](https://github.com/pointfreeco/swift-dependencies/assets/48778177/46283384-40e7-4938-bdcf-f49b7362ac81)

**Reply by mbrandonw** [reply]
That's good to know, but the other bug is also a problem. Currently you cannot use property wrappers inside `@Observable`. You can see that in the error in @hadiidbouk's screenshot:

> 🛑 Property wrapper cannot be applied to a computed property

…as well as the code sample I posted.

So that too needs to be fixed before you can use a bare `@Dependency` inside an `@Observable`. But also there is no harm in marking the dependency as `@ObservationIgnored` since it really does not need to be observed.

**Reply by roundfly** [reply]
If I understood correctly, once [this](https://github.com/hborla/swift-evolution/blob/init-accessors/proposals/NNNN-init-accessors.md) pitch lands (along with a [follow up PR](https://forums.swift.org/t/second-review-se-0395-observability/65261/23) for Observability) the limitation of having to provide default values will be lifted.
